Add default cf push functionality

JordanICollier opened the pull request:
We have implemented default cf push functionality. We added a unit test for the cli connection and verified it worked. We added feature test for the run command.

cc: @trayo

trayo committed:
Add default cf push functionality

[#106304374]

Signed-off-by: Jordan Collier <jordanicollier@gmail.com>

JordanICollier commented:
Would you mind elaborating on the idea of implementing `appNotFound()` within `rewind.Actions`? One of our concerns about doing this is, it does not seem like there is a way to prevent it from running the other defined actions. For example, we cannot put an if else statement within the struct itself. We could make it do a normal push if AppNotFound, but it will carry on and also execute zero down time push, because there is not way to prevent the rest of the actions from executing.

JordanICollier commented:
So here is our implementation of putting what you suggested within rewind actions. With this version you have to return an error in order to not run the rest of the actions. Another thing we had to change within the functionality is the `fatalif()` command no longer returns an exit 1because it will cause the tests fail.
![screen shot 2015-11-04 at 10 30 23 am](https://cloud.githubusercontent.com/assets/6378836/10946370/29b1c5a2-82df-11e5-8477-107a1b422474.png)

This is the other way we found of implementing the rewind actions functionality. With this functionality it will push a new app and then also do a zdd-push immediately after. i.e. it will push an app twice when it is a new app.
![screen shot 2015-11-04 at 11 50 14 am](https://cloud.githubusercontent.com/assets/6378836/10948315/48d31e30-82ea-11e5-9900-91da75291a59.png)

Here is the resulting output of the double push:

```
my-new-app not found! Using cf push.

Creating app my-new-app in org IS-COMPOZED-DEPLOYER / space DEV as tminus...
OK

Creating route my-new-app.apps.com...
OK

Binding my-new-app.apps.com to my-new-app...
OK

Uploading my-new-app...
Uploading app files from: /Users/localadmin/workspace/my-new-app
Uploading 552, 3 files
Done uploading
OK

Starting app my-new-app in org IS-COMPOZED-DEPLOYER / space DEV as tminus...
-----> Downloaded app package (4.0K)
-----> Using root folder
-----> Copying project files into public/
-----> Setting up nginx

-----> Uploading droplet (3.3M)

1 of 1 instances running

App started

#### This is where it starts the push-zdd

OK

App my-new-app was started using this command `sh boot.sh`

Showing health and status for app my-new-app in org IS-COMPOZED-DEPLOYER / space DEV as tminus...
OK

requested state: started
instances: 1/1
usage: 1G x 1 instances
urls: my-new-app.apps.com
last uploaded: Wed Nov 4 18:55:57 UTC 2015
stack: cflinuxfs2
buildpack: Static file

     state     since                    cpu    memory       disk         details
     #0   running   2015-11-04 11:56:09 AM   0.0%   4.5M of 1G   8.6M of 1G
     Renaming app my-new-app to my-new-app-venerable in org IS-COMPOZED-DEPLOYER / space DEV as tminus...
     OK
     Creating app my-new-app in org IS-COMPOZED-DEPLOYER / space DEV as tminus...
     OK

     Using route my-new-app.apps.com
     Binding my-new-app.apps.com to my-new-app...
     OK

     Uploading my-new-app...
     Uploading app files from: /Users/localadmin/workspace/my-new-app
     Uploading 552, 3 files
     Done uploading
     OK

     Starting app my-new-app in org IS-COMPOZED-DEPLOYER / space DEV as tminus...
     -----> Downloaded app package (4.0K)
     -----> Using root folder
     -----> Copying project files into public/
     -----> Setting up nginx

     -----> Uploading droplet (3.3M)

     1 of 1 instances running

     App started


     OK

     App my-new-app was started using this command `sh boot.sh`

     Showing health and status for app my-new-app in org IS-COMPOZED-DEPLOYER / space DEV as tminus...
     OK

     requested state: started
     instances: 1/1
     usage: 1G x 1 instances
     urls: my-new-app.apps.com
     last uploaded: Wed Nov 4 18:56:22 UTC 2015
     stack: cflinuxfs2
     buildpack: Static file

          state     since                    cpu    memory       disk         details
          #0   running   2015-11-04 11:56:34 AM   0.0%   4.5M of 1G   8.6M of 1G
          Deleting app my-new-app-venerable in org IS-COMPOZED-DEPLOYER / space DEV as tminus...
          OK

          A new version of your application has successfully been pushed!
```

xchapter7x commented:
Hi,

Sorry, i was def not too clear on the direction i was suggesting.

What I am suggesting, and hoping to hear your collective thoughts on is
this.

We might consider simply creating a single 'push' rewind.Action. Since this
needs to happen regardless which state the app is in. Then employing a
decorator to either add the rename and delete steps or not, based on if it
is a new app or not.

in the future we could simply add more decorator layers if we needed to add
more logic or steps or whatever the case may be.

I thew together this example in go playground to convey my intent a bit
more. It def doesn't compile, so treat it as pseudo code.
https://play.golang.org/p/Y-eyUYfXKe

The real question is, will this approach help us, in terms of clarity,
quality and maintainability or will it make things more confusing? Thoughts?

BTW, I think conversations like this are really important and healthy. I
hope the feedback is not taken as me being overly critical. Keep up the
great work and keep contributing.

Regards,

John Calabrese

2015-11-04 14:00 GMT-05:00 Jordan I Collier notifications@github.com:

> So here is our implementation of putting what you suggested within rewind
> actions. With this version you have to return an error in order to not run
> the rest of the actions. Another thing we had to change within the
> functionality is the fatalif() command no longer returns an exit 1because
> it will cause the tests fail.
> [image: screen shot 2015-11-04 at 10 30 23 am]
> https://cloud.githubusercontent.com/assets/6378836/10946370/29b1c5a2-82df-11e5-8477-107a1b422474.png
> 
> This is the other way we found of implementing the rewind actions
> functionality. With this functionality it will push a new app and then also
> do a zdd-push immediately after. i.e. it will push an app twice when it is
> a new app.
> [image: screen shot 2015-11-04 at 11 50 14 am]
> https://cloud.githubusercontent.com/assets/6378836/10948315/48d31e30-82ea-11e5-9900-91da75291a59.png
> 
> Here is the resulting output of the double push:
> 
> my-new-app not found! Using cf push.
> 
> Creating app my-new-app in org IS-COMPOZED-DEPLOYER / space DEV as tminus...
> OK
> 
> Creating route my-new-app.apps.com...
> OK
> 
> Binding my-new-app.apps.com to my-new-app...
> OK
> 
> Uploading my-new-app...
> Uploading app files from: /Users/localadmin/workspace/my-new-app
> Uploading 552, 3 files
> Done uploading
> OK
> 
> Starting app my-new-app in org IS-COMPOZED-DEPLOYER / space DEV as tminus...
> -----> Downloaded app package (4.0K)
> -----> Using root folder
> -----> Copying project files into public/
> -----> Setting up nginx
> 
> -----> Uploading droplet (3.3M)
> 
> 1 of 1 instances running
> 
> App started
> 
> #### This is where it starts the push-zdd
> 
> OK
> 
> App my-new-app was started using this command `sh boot.sh`
> 
> Showing health and status for app my-new-app in org IS-COMPOZED-DEPLOYER / space DEV as tminus...
> OK
> 
> requested state: started
> instances: 1/1
> usage: 1G x 1 instances
> urls: my-new-app.apps.com
> last uploaded: Wed Nov 4 18:55:57 UTC 2015
> stack: cflinuxfs2
> buildpack: Static file
> 
> ```
>  state     since                    cpu    memory       disk         details
>  #0   running   2015-11-04 11:56:09 AM   0.0%   4.5M of 1G   8.6M of 1G
>  Renaming app my-new-app to my-new-app-venerable in org IS-COMPOZED-DEPLOYER / space DEV as tminus...
>  OK
>  Creating app my-new-app in org IS-COMPOZED-DEPLOYER / space DEV as tminus...
>  OK
> 
>  Using route my-new-app.apps.com
>  Binding my-new-app.apps.com to my-new-app...
>  OK
> 
>  Uploading my-new-app...
>  Uploading app files from: /Users/localadmin/workspace/my-new-app
>  Uploading 552, 3 files
>  Done uploading
>  OK
> 
>  Starting app my-new-app in org IS-COMPOZED-DEPLOYER / space DEV as tminus...
>  -----> Downloaded app package (4.0K)
>  -----> Using root folder
>  -----> Copying project files into public/
>  -----> Setting up nginx
> 
>  -----> Uploading droplet (3.3M)
> 
>  1 of 1 instances running
> 
>  App started
> 
> 
>  OK
> 
>  App my-new-app was started using this command `sh boot.sh`
> 
>  Showing health and status for app my-new-app in org IS-COMPOZED-DEPLOYER / space DEV as tminus...
>  OK
> 
>  requested state: started
>  instances: 1/1
>  usage: 1G x 1 instances
>  urls: my-new-app.apps.com
>  last uploaded: Wed Nov 4 18:56:22 UTC 2015
>  stack: cflinuxfs2
>  buildpack: Static file
> 
>       state     since                    cpu    memory       disk         details
>       #0   running   2015-11-04 11:56:34 AM   0.0%   4.5M of 1G   8.6M of 1G
>       Deleting app my-new-app-venerable in org IS-COMPOZED-DEPLOYER / space DEV as tminus...
>       OK
> 
>       A new version of your application has successfully been pushed!
> ```
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/xchapter7x/autopilot/pull/3#issuecomment-153831226.

JordanICollier commented:
So is the push action now responsible for renaming and deleting? Would it be better if we extract all of this rewind functionality out of the run function to thin out the cruft?( BTW cruft is now our favorite word, so thank you for that)

Yes we agree that these conversations are invaluable, we do not view them as being over critical.

xchapter7x commented:
Right, I think we are saying the same thing. Break all of the actions out.
Make a very thin Run method. And compose the different possible
rewind.actions into the the action list.

We might do this composition via a decorator, as in my pseudo code example,
but that is really just an implementation detail than a requirement.

So general goal / direction we seem to have consensus on is a thin Run()
method with a composable rewind.actions list?

-john

On Wed, Nov 4, 2015, 5:25 PM Jordan I Collier notifications@github.com
wrote:

> So is the push action now responsible for renaming and deleting? Would it
> be better if we extract all of this rewind functionality out of the run
> function to thin out the cruft?( BTW cruft is now our favorite word, so
> thank you for that)
> 
> Yes we agree that these conversations are invaluable, we do not view them
> as being over critical.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/xchapter7x/autopilot/pull/3#issuecomment-153888848.

trayo commented:
Yeah, I think we were thinking the same thing. We looked over your example a bit more and add some changes to understand more what that direction would look like.

We will continue pressing forward and see what we can come up with.

What do you think about separating out the CliCommand functions at the bottom ([line 104 to and down](https://github.com/xchapter7x/autopilot/blob/master/autopilot.go#L104)) out into it's own plugin. The tests around those functions would then be separated out from the `autopilot_test.go` and the `autopilot_test.go` would be responsible for higher level feature tests around the `run()` method.

JordanICollier committed:
extract applicationrepo functionality into its own package

[#106304374]

Signed-off-by: Travis Yoder <trayoda@gmail.com>

trayo committed:
Add exported ActionList array to test number of actions

this is used to verify the number of actions called when pushing
a new app or an old app

[#106304374]

Signed-off-by: Jordan Collier <jordanicollier@gmail.com>

JordanICollier committed:
extract call to ListApplications into a function

[#106304374]

Signed-off-by: Travis Yoder <trayoda@gmail.com>

trayo committed:
change appNotFound to appExists

[#106304374]

Signed-off-by: Jordan Collier <jordanicollier@gmail.com>

JordanICollier committed:
implement functions to get actions for pushing

functions added:
+ getPushAction(argList []string) rewind.Action
+ addReversePrevious(action *rewind.Action)
+ getRenameAction() rewind.Action
+ getDeleteAction() rewind.Action

[#106304374]

Signed-off-by: Travis Yoder <trayoda@gmail.com>

trayo committed:
add appName and venerableAppName as properties on AutoPilotPlugin

[#106304374]

Signed-off-by: Jordan Collier <jordanicollier@gmail.com>

JordanICollier committed:
created getActions function responsible for building ActionList

functions added:
+ getAction(argList []string) []rewind.Action

[#106304374]

Signed-off-by: Travis Yoder <trayoda@gmail.com>

trayo committed:
refactored run to getActions and execute those actions

[#106304374]

Signed-off-by: Jordan Collier <jordanicollier@gmail.com>

JordanICollier committed:
move anonymous functions to bottom of file for better clarity

[#106304374]

Signed-off-by: Travis Yoder <trayoda@gmail.com>

JordanICollier commented:
@xchapter7x So we have sent quite a few commits, did a bunch of refactoring to implement your ideas. Let us know what you think.

xchapter7x commented on autopilot_test.go:
```diff
@@ -155,4 +49,25 @@ var _ = Describe("Command Syntax", func() {
+			autopilotPlugin.Run(cliConn, []string{"push-zdd", "myapp"})
+		})
+
+		Expect(len(ActionList)).To(Equal(3))
```
rather than get the length of this ActionList array, which requires us to make it exported, you might consider just leveraging the cli plugin fakes.

`CliCommandCallCount` will get you the call count
https://github.com/cloudfoundry/cli/blob/master/plugin/fakes/fake_cli_connection.go#L267

`CliCommandArgsForCall` will get you the actual argument sets called.
https://github.com/cloudfoundry/cli/blob/master/plugin/fakes/fake_cli_connection.go#L273

xchapter7x commented on autopilot_test.go:
```diff
@@ -155,4 +49,25 @@ var _ = Describe("Command Syntax", func() {
+		})
+
+		Expect(len(ActionList)).To(Equal(1))
+		Expect(output).ToNot(ContainElement(ContainSubstring("using zero-downtime-deployment")))
```
similar to above,
we dont need to test for a specific string being printed. CLI output isnt really our concern in this particular behavior we are testing. It also might change, which would fail our test.

consider using:
`CliCommandCallCount` will get you the call count
https://github.com/cloudfoundry/cli/blob/master/plugin/fakes/fake_cli_connection.go#L267

`CliCommandArgsForCall` will get you the actual argument sets called.
https://github.com/cloudfoundry/cli/blob/master/plugin/fakes/fake_cli_connection.go#L273

xchapter7x commented on autopilot.go:
```diff
@@ -4,67 +4,44 @@ import (
-		fmt.Fprintln(os.Stdout, "error:", err)
-		os.Exit(1)
-	}
+var ActionList []rewind.Action
```
as stated in other comments, we probably don't want to export this as we can more accurately test using provided fakes.

xchapter7x commented on application_repo/application_repo.go:
```diff
@@ -0,0 +1,45 @@
+}
+
+//ListApplications - list applications on cf
+func (repo *ApplicationRepo) ListApplications() error {
```
i think these existed previously, but we might consider removing them in favor of 
`GetApps()` in our repo.conn object
https://github.com/cloudfoundry/cli/blob/master/plugin/cli_connection.go#L312

xchapter7x commented:
great job guys.
just a few feedback points above.

xchapter7x commented:
going to pull this in and make suggested as a follow on commit

trayo commented:
Thanks @xchapter7x ! :rocket:
